qiskit_experiments/calibration/experiments/rabi.py

```diff
@@ -0,0 +1,279 @@
+from qiskit.providers.options import Options
+
+from qiskit_experiments.analysis import (
+    CurveAnalysis,
+    CurveAnalysisResult,
+    SeriesDef,
+    fit_function,
+    get_opt_value,
+    get_opt_error,
+)
+from qiskit_experiments.base_experiment import BaseExperiment
+from qiskit_experiments.data_processing.processor_library import get_to_signal_processor
+
+
+class RabiAnalysis(CurveAnalysis):
+    r"""Rabi analysis class based on a fit to a cosine function.
+
+    Analyse a Rabi experiment by fitting it to a cosine function
+
+    .. math::
+        y = a \cos\left(2 \pi {\rm freq} x + {\rm phase}\right) + b
+
+    Fit Parameters
+        - :math:`a`: Amplitude of the oscillation.
+        - :math:`b`: Base line.
+        - :math:`{\rm freq}`: Frequency of the oscillation. This is the fit parameter of interest.
+        - :math:`{\rm phase}`: Phase of the oscillation.
+
+    Initial Guesses
+        - :math:`a`: The maximum y value less the minimum y value.
+        - :math:`b`: The average of the data.
+        - :math:`{\rm freq}`: The frequency with the highest power spectral density.
+        - :math:`{\rm phase}`: Zero.
+
+    Bounds
+        - :math:`a`: [-2, 2] scaled to the maximum signal value.
+        - :math:`b`: [-1, 1] scaled to the maximum signal value.
+        - :math:`{\rm freq}`: [0, inf].
+        - :math:`{\rm phase}`: [-pi, pi].
+    """
+
+    __series__ = [
+        SeriesDef(
+            fit_func=lambda x, a, freq, phase, b: fit_function.cos(
+                x, amp=a, freq=freq, phase=phase, baseline=b
+            ),
+            plot_color="blue",
+        )
+    ]
+
+    @classmethod
+    def _default_options(cls):
+        """Return the default analysis options.
+
+        See :meth:`~qiskit_experiment.analysis.CurveAnalysis._default_options` for
+        descriptions of analysis options.
+        """
+        default_options = super()._default_options()
+        default_options.p0 = {"a": None, "freq": None, "phase": None, "b": None}
+        default_options.bounds = {"a": None, "freq": None, "phase": None, "b": None}
+        default_options.fit_reports = {"freq": "frequency"}
+        default_options.normalization = True
+        default_options.xlabel = "Amplitude"
+        default_options.ylabel = "Signal (arb. units)"
+
+        return default_options
+
+    def _setup_fitting(self, **options) -> Union[Dict[str, Any], List[Dict[str, Any]]]:
+        """Fitter options."""
+        user_p0 = self._get_option("p0")
+        user_bounds = self._get_option("bounds")
+
+        max_abs_y = np.max(np.abs(self._y_values))
+
+        # Use a fast Fourier transform to guess the frequency.
+        fft = np.abs(np.fft.fft(self._y_values))
+        damp = self._x_values[1] - self._x_values[0]
+        freqs = np.linspace(0.0, 1.0 / (2.0 * damp), len(fft // 2))
+
+        a_guess = np.max(self._y_values) - np.min(self._y_values)
+        f_guess = freqs[np.argmax(fft)]
+        p_guess = 0
```

Comment:
How robust is this initial guess? In IQ measurement the initial phase will become quite arbitrary. Perhaps it's better to prepare `fit_oiption` for `p_guesses = [0, np.pi / 4, np.pi / 2, 3 / 4 * np.pi]`.

Reply:
For a Rabi the phase should not be arbitrary. I expect only 0 or pi even with IQ data as an amplitude of 0 should always correspond to an extrema. That being said, adding a few initial guesses is not a bad idea. See here: 84e3c36